packages/grid/_modules_/grid/hooks/core/useGridApiInitialization.ts

```diff
@@ -11,11 +12,45 @@ const isSyntheticEvent = (event: any): event is React.SyntheticEvent => {
   return event.isPropagationStopped !== undefined;
 };
 
+const wrapPublicApi = (publicApi: GridApi) => {
+  const privateOnlyApi = {
+    getPublicApi: () => publicApi,
+    registerMethod: (methodName: string, isPublic: boolean, method: (...args: any[]) => any) => {
+      if (isPublic) {
+        publicApi[methodName] = method;
+      } else {
+        privateOnlyApi[methodName] = method;
+      }
+    },
+  } as Omit<GridPrivateApi, keyof GridApi>;
+
+  const handler: ProxyHandler<GridApi> = {
+    get: (obj, prop) => {
+      if (obj[prop] !== undefined) {
+        return obj[prop];
+      }
+
+      return privateOnlyApi[prop];
+    },
+    set: (obj, prop, value) => {
+      obj[prop] = value;
+      return true;
+    },
+  };
+
+  return new Proxy(publicApi, handler) as GridPrivateApi;
+};
+
 export function useGridApiInitialization(
-  apiRef: GridApiRef,
+  publicApiRef: GridApiRef,
   props: Pick<GridComponentProps, 'signature'>,
-): void {
-  const logger = useGridLogger(apiRef, 'useApi');
+): GridPrivateApiRef {
+  const privateApiRef = React.useRef() as GridPrivateApiRef;
+  if (!privateApiRef.current) {
+    privateApiRef.current = wrapPublicApi(publicApiRef.current);
```

Comment:
One limitation here is that we must never change the reference of `apiRef.current` otherwise the private will still track the previous one.
We are never doing it and I think we could have other issues if we do.